### PR TITLE
Add an endpoint to the CMS that returns some stats for the Admin Question Dashboard

### PR DESCRIPTION
**Pull request by cissyxyu**
## WHAT
As part of building out a new Admin Question Dashboard, I want to add an endpoint to the CMS server that returns some calculated data about question health.

## WHY
This data is important for admin to know, and we need to put the endpoint in CMS because the LMS does not store response authors/matched/unmatched data.

## HOW
Add the endpoint in CMS and make sure it's working and the `lib` path is loaded by the server.

### Screenshots
(If applicable. Also, please censor any sensitive data)

### Notion Card Links
https://www.notion.so/quill/Change-CMS-to-provide-correct-back-end-data-f95e6cd0c4f94e4aa0f39f1b4d7f2e02

PR Checklist | Your Answer
------------ | -------------
Have you added and/or updated tests? |  YES
Have you deployed to Staging? | Not yet - deploying now!
Self-Review: Have you done an initial self-review of the code below on Github? | Yes
Design Review: If applicable, have you compared the coded design to the mockups? | N/A

### CODE REVIEW

File: services/QuillCMS/app/controllers/responses_controller.rb [hunk 1]
```diff
@@ -94,6 +94,10 @@ def grade_breakdown
     render json: optimality_counts_of_question(params[:question_uid])
   end
 
+  def question_dashboard
+    render json: question_dashboard_data(params[:question_uid])
+  end
+
   def incorrect_sequences
     render json: IncorrectSequenceCalculator.incorrect_sequences_for_question(params[:question_uid])
```
Comment (happythenewsad) [inline services/QuillCMS/app/controllers/responses_controller.rb:103]: IncorrectSequenceCalculator, which lives in lib/modules, is being loaded here without the need of your modification to `config/application.rb`. Do you need  the modification in `config/application.rb`?

File: services/QuillCMS/lib/modules/response_aggregator.rb [hunk 2]
```diff
@@ -32,6 +42,17 @@ def optimality_counts_of_question(question_uid)
...
   end
 
+  def question_dashboard_data(question_uid)
+    total_number_of_responses = Response.where(question_uid: question_uid).sum('count')
+    common_unmatched = Response.where(question_uid: question_uid, parent_id: nil, optimal: nil).where('count >= 10').sum('count')
+    algorithm_matched = Response.where(question_uid: question_uid, author: DASHBOARD_ALGOS).sum('count')
+
+    {
+      percent_common_unmatched: (common_unmatched.to_f / (total_number_of_responses.nonzero? || 1)) * 100,
+      percent_specified_algos: (algorithm_matched.to_f / (total_number_of_responses.nonzero? || 1)) * 100
```
Comment (happythenewsad) [inline services/QuillCMS/lib/modules/response_aggregator.rb]: Interesting - `nonzero?` is the only Rails API I've encountered that breaks the convention that methods ending in ? return booleans.

Comment (brendanshean) [reply services/QuillCMS/lib/modules/response_aggregator.rb]: `File#size?` is another gotcha.

Comment (brendanshean) [reply services/QuillCMS/lib/modules/response_aggregator.rb]: I think what's most confusing is `zero?` works as expected.

File: services/QuillCMS/spec/controllers/responses_controller_spec.rb [hunk 3]
```diff
@@ -215,4 +215,25 @@ def filter_responses_with_parent_id(array)
...
     end
   end
+
+  describe 'question_dashboard_data' do
+    before do
+      allow_any_instance_of(Response).to receive(:create_index_in_elastic_search)
+    end
+
+    let!(:question_uid) { SecureRandom.uuid}
+    let!(:response_1) { create(:response, question_uid: question_uid, parent_id: nil, optimal: nil, count: 20)}
```
Comment (brendanshean) [inline services/QuillCMS/spec/controllers/responses_controller_spec.rb]: I don't think it's necessary to include `parent_id: nil` as the factory doesn't mention that field.

File: services/QuillCMS/spec/lib/modules/response_aggregator_spec.rb [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+RSpec.describe ResponseAggregator do
+  include ResponseAggregator
+
+  describe "#question_dashboard_data" do
+    before do
+      allow_any_instance_of(Response).to receive(:create_index_in_elastic_search)
+    end
+
+    let!(:question_uid) { SecureRandom.uuid}
+    let!(:response_1) { create(:response, question_uid: question_uid, parent_id: nil, optimal: nil, count: 20)}
```
Comment (brendanshean) [inline services/QuillCMS/spec/lib/modules/response_aggregator_spec.rb]: I don't think it's necessary to include `parent_id: nil` as the factory doesn't mention that field.

File: services/QuillCMS/lib/modules/response_aggregator.rb (hunk 2 above)
Comment (brendanshean) [reply services/QuillCMS/lib/modules/response_aggregator.rb]: Actually, I think using `zero?` would be more clear for both `percent_common_umatched` and `percent_specified_algos`.  For example (assuming refactor from module into an object like `AdminQuestionDashboard`):

```
def percent_common_unmatched
  num_responses.zero? ? 0 : (100.0 * common_unmatched / num_of_responses)
end
```

**Review by brendanshean — COMMENTED** [review]
I think this would be a good opportunity to move away from the `lib/modules` pattern by introducing a new object under `app/models` called `AdminQuestionDashboard` with one public method `health`.  This would do a few things:

1) Adhere to single responsibility as well as encapsulation. (adding to ResponseAggregator violates both)
2) Cut down on parameter passing
3) Give more natural specs.

I'm happy to pair on this.

File: services/QuillCMS/app/controllers/responses_controller.rb (hunk 1 above)
Comment (cissyxyu) [reply services/QuillCMS/app/controllers/responses_controller.rb:103]: I changed the path that this file lives in, so removed the autoload modification line!

**Comment by cissyxyu** [comment]
> I think this would be a good opportunity to move away from the `lib/modules` pattern by introducing a new object under `app/models` called `AdminQuestionDashboard` with one public method `health`. This would do a few things:
> 
> 1. Adhere to single responsibility as well as encapsulation. (adding to ResponseAggregator violates both)
> 2. Cut down on parameter passing
> 3. Give more natural specs.
> 
> I'm happy to pair on this.

Changed the module pattern to be what you said -- made a new model in app/models!

**Review by happythenewsad — APPROVED** [review]

**Review by brendanshean — APPROVED** [review]

**Pull request merged**